h/api/logic.py

```diff
@@ -66,13 +63,11 @@ def update_annotation(annotation, fields, has_admin_permission):
         fields.pop(field, None)
 
     # If the user is changing access permissions, check if it's allowed.
-    changing_permissions = (
-        'permissions' in fields and
-        fields['permissions'] != annotation.get('permissions', {})
-    )
-    if changing_permissions and not has_admin_permission:
-        raise RuntimeError(
-            _('Not authorized to change annotation permissions.'), 401)
+    permissions = annotation.get('permissions', {})
+    if fields.get('permissions', {}) != permissions:
+        if user not in permissions.get('admin', []):
+            raise RuntimeError(
+                _('Not authorized to change annotation permissions.'), 401)
```

Comment:
BTW, instead of this weird leakage of HTTP codes upward, these functions, if we really want to keep this logic layer, could throw custom exception classes that we could just catch as exception views and turn into HTTP exceptions.